Support for train data augmentation

groar opened the pull request:
Added data augmentation to classification dataset creation. Only the train set is augmented at db creation with new images obtained by a combination of the following transformations : 
- Rotation 
- Translation
- Contrast modulation
- Hue modulation

There are a few parameters for each transformation, including the probability for each image of being augmented. 

<img width="300" alt="Dataset augmentation" src="https://cloud.githubusercontent.com/assets/5939621/10099739/bd8d96ec-638c-11e5-8654-ccabdb2dad87.png">

For now, only added tests to tools/create_db, but will add a few more regarding task creation and utils/image. Edit: done. 

Translating is quite slow, so I should check if the way I do it can be improved. 

### TODO before merge

- [ ] Fix hue modulation and translation
- [ ] Add the possibility of setting the number of images generated out of each training sample
- [ ] Fix the UI

lukeyeager commented:
Wow this is awesome, great work! My little smoketests all look good so far, I'll try to dig into the code some more today and see if I have any feedback for you.

Now is when I really wish we had #21 so I could see this stuff in action ...

lukeyeager commented on digits/dataset/images/classification/views.py:
```diff
@@ -88,35 +88,70 @@ def from_folders(job, form):
+                    encoding     = encoding,
+                    compression  = compression,
+                    labels_file  = job.labels_file,
+                    augmentation = {},
```
You can just do here what you did below at line 171. I think it's better to keep the number of changed lines small than preserve indentation. I sort of regret hand-aligning indentation in general.

lukeyeager commented on digits/dataset/tasks/create_db.py:
```diff
@@ -11,7 +11,7 @@
 
 # NOTE: Increment this everytime the pickled version changes
-PICKLE_VERSION = 3
+PICKLE_VERSION = 4
```
I don't think we need to increment the version here. We can just add this snippet to the end of `__set_state__()`:
```python
if not hasattr(self, 'augmentation'):
    self.augmentation = None
```

lukeyeager commented on digits/templates/datasets/images/classification/show.html:
```diff
@@ -101,6 +101,17 @@
     <img src="{{url_for('serve_file', path=task.path('mean.jpg', relative=True))}}" />
     {% endif %}
+    <br/>
+    <b>Augmentations:</b>
```
This should only be shown if some augmentations exist. Otherwise you end up with this:

![augmentations-none-display-issue](https://cloud.githubusercontent.com/assets/687269/10109099/c7aabbea-6379-11e5-9042-d44d16a7e798.jpg)

lukeyeager commented on digits/utils/image.py:
```diff
@@ -378,3 +504,75 @@ def get_color_map(name):
+def rgb_to_hsv(rgb):
+    # Translated from source of colorsys.rgb_to_hsv
+    # r,g,b should be a numpy arrays with values between 0 and 255
+    # rgb_to_hsv returns an array of floats between 0.0 and 1.0.
```
It looks like you're getting this code from [this SO post](http://stackoverflow.com/a/7274986/2404152). SO licenses user contributions with a viral license.

> user contributions licensed under cc by-sa 3.0 with attribution required

http://creativecommons.org/licenses/by-sa/3.0/

Unfortunately, that means you're going to need to re-implement this yourself or get permission from the original author to use this code :/

lukeyeager commented:
Really great stuff. I've mentioned a few nitpicks above (and the licensing thing - hopefully you can find a good solution for that). Otherwise, this LGTM!

@jmancewicz, any thoughts?

groar commented:
Wow, the SO license is a good point I should have checked ! I replaced those functions with matplotlib ones. It seems they are already implemented in Pillow (see https://github.com/python-pillow/Pillow/pull/816) but I couldn't figure out for sure. 

I also ignored hue modulation in the grayscale case since it does not make sense (we should do something better, like disabling the checkbox in the "new dataset" view). 

I think I have corrected all the things you've mentioned. Tell me if you see something else.

lukeyeager commented:
@groar, Awesome, thanks for the changes! Can you squash this down and remove the extra commits?

@gheinrich, @jmancewicz - I'd feel better if one of y'all took a look at this too before merging. I don't have time to test this exhaustively, but it looks pretty good to me.

gheinrich commented on digits/dataset/images/classification/test_views.py:
```diff
@@ -66,7 +66,7 @@ class BaseViewsTestWithImageset(BaseViewsTest):
     IMAGE_CHANNELS  = 3
     BACKEND         = 'lmdb'
-    COMPRESSION     = 'none'
+    COMPRESSION     = 'none' 
```
I think trailing spaces should be avoided if possible (especially here since there is no functional change to this line)

lukeyeager replied:
Yeah, there's a lot of this. I use [this plugin](https://github.com/bronson/vim-trailing-whitespace) (shameless plug), which makes the whitespace issues super obvious.

gheinrich replied:
My text editor (visual slick) has an option to automatically remove trailing spaces.

groar replied:
Ok I finally installed a sublime plugin to highlight those trailing spaces! I should be able not to add any new trailing space at least.

gheinrich commented on digits/dataset/tasks/create_db.py:
```diff
@@ -91,13 +93,15 @@ def __setstate__(self, state):
         if not hasattr(self, 'compression') or self.compression is None:
             self.compression = 'none'
+        if not hasattr(self, 'augmentation'):
+            self.augmentation = None
```
maybe you need to do:
`self.augmentation = {}`
I had an error when loading an older dataset:
```
  File "/fast-scratch/gheinrich/ws/digits/digits/templates/datasets/images/classification/show.html", line 105, in template
    {% if task.augmentation|length != 0 %}
TypeError: object of type 'NoneType' has no len()
```
I think this is because self.augmentation is set to None here.

groar replied:
Probably yes, will fix that.

gheinrich commented:
I have a question about the terminology here: in my understanding 'data augmentation' is a method to artificially create *many* unseen training examples by performing transformations on existing samples from the dataset. Am I understanding correctly that this PR does not actually increase the number of training samples? In this case is it technically correct to call this 'data augmentation'? I am asking candidly as I am not sure my understanding of the term is correct.

Either way, would it be useful to add an option to let the user specify a measure of how many images may be generated out of each training sample in the original dataset?

gheinrich commented on digits/dataset/images/classification/test_views.py:
```diff
@@ -442,6 +459,9 @@ class TestCreatedWide(TestCreated):
 
+class TestCreatedAugmentation(TestCreated):
+    IMAGE_CHANNELS = 1
+
```
maybe add a test with IMAGE_CHANNELS=3 so you can test hue augmentation?

groar replied:
Hum yeah I don't know why I added those lines, they are testing nothing. The real tests are done in `TestAugmentedDataset`. I will test the different augmentations yes.

gheinrich commented on digits/dataset/images/classification/views.py:
```diff
@@ -167,6 +203,37 @@ def from_files(job, form):
+
+    if form.has_augmentation_contrast.data:
+        augmentation['contrast'] = {
+            'strength': form.augmentation_contrast_strength.data,
```
I might be missing something but I think this field does not exist? Actually when trying to create a dataset from text files I get:
```
  File "/fast-scratch/gheinrich/ws/digits/digits/dataset/images/classification/views.py", line 211, in from_files
    'strength': form.augmentation_contrast_strength.data,
AttributeError: 'ImageClassificationDatasetForm' object has no attribute 'augmentation_contrast_strength'
```
Besides this code looks similar to lines 91 to 192 but is not exactly the same (I think the correct version is in the lines 91 to 192). Can you move this to a function and remove the duplicate code?

groar replied:
This is a residual from a previous version, well spotted. I'll remove it.

gheinrich commented on digits/dataset/images/classification/forms.py:
```diff
@@ -38,6 +38,158 @@ def validate_backend(form, field):
+                ]
+            )
+ 
+    augmentation_hue_angle = utils.forms.FloatField(u'angle',
```
is this really needed?

groar replied:
What do you mean? The `has_augmentation_hue` field?

lukeyeager replied:
I think he means the `augmentation_hue_angle` field. Why does a hue have an angle?

gheinrich replied:
I meant that the field `augmentation_hue_angle` appears to be unused (`augmentation_hue_angle_min` and `augmentation_hue_angle_max` are used instead).

gheinrich commented on digits/dataset/images/classification/views.py:
```diff
@@ -167,6 +203,37 @@ def from_files(job, form):
+
+    if form.has_augmentation_hue.data and job.image_dims[2] == 3:
+        augmentation['hue'] = {
+            'angle': form.augmentation_hue_angle.data,
```
do you mean to use `hue_angle_min` and `hue_angle_max`? The field `hue_angle` doesn't seem to be rendered in the HTML template...

groar replied:
Yeah this is related to the other remark you made.

groar commented:
> Would it be useful to add an option to let the user specify a measure of how many images may be  generated out of each training sample in the original dataset? (as opposed to just the probability that each transformation is applied)

@gheinrich Yes this is indeed important. We should have both a given probability and the possibility of getting several images out of an original training sample for a given transformation. I will do that as soon as I get some time.

lukeyeager commented:
@groar, it sounds like you have some more things to do before you feel like this is ready to merge.

Can you add a TODO list to the original post (like #209, #226, #268) so we can see the progress?

jmancewicz commented:
First off, this is awesome... but I have a few initial thoughts

A little thing, just about the code.  Can you try to not leave trailing whitespaces.  Future changes, to whitespace can cause merge and ownership issues. It's a good practice to strip them before commit.

I think it makes sense to move the transformation matrix operations together.  I suggest moving 'Perform translation on images' beneath 'Perform rotations on images'.  Also, and this is being picky, but translation is singular and rotations is plural.  Actually, I'd skip changing that and make the phrases more succinct.  Perhaps...

![augment](https://cloud.githubusercontent.com/assets/13259615/10178850/2c0944cc-66b7-11e5-81df-2647b0767485.png)

or

![augment2](https://cloud.githubusercontent.com/assets/13259615/10178852/34da2472-66b7-11e5-9e9e-dea5047a56fb.png)

That might be too vague, but I doubt it. The tooltips make everything clear.

Generally, field labels are capitalized [Image size] (some are title style [Resize Transformation], and maybe should be changed), but with the exception of the 'dx' and 'dy', I recommend capitalization of these labels. 

I'm conflicted about the unit of dx and dy.  I'm fine with it being in normalized image units, but wonder if % would make it clearer what's going to happen to the images.

I'll dig into the code now.

jmancewicz commented:
Would it be better to not use clamp to edge wrap mode?  Is it better to use a solid color?

![translate](https://cloud.githubusercontent.com/assets/13259615/10181272/f876503c-66ca-11e5-9a1f-d27c0003debd.png)

jmancewicz commented:
Also, the hue modulation does not appear to be modulation.  It looks like it's setting the hue to a constant for each new image.  That might be fine, but I expected it to rotate the hue.

by changing the code from

    hsv[:,:, 0] = angle / 360.0
to
    hsv[:,:, 0] += angle / 360.0
    hsv[:,:, 0] %= 1
or if you wanted to extend the range from [0,360] to [-360,360], then 
    hsv[:,:, 0] += angle / 360.0 + 1.0
    hsv[:,:, 0] %= 1
should do.  That's all assuming that you meant to shift the hue.  If the intent was to set the hue, then maybe we shouldn't call it modulation.  I'm not familiar with how hue affects training.  It almost seems like a rgb mult would do better for simulating varying lighting conditions.

jmancewicz commented:
Yes, the translation is very slow.  I wonder if it is from the clamp to edge wrap mode.  I expect this could be made much faster using OpenGL.

groar commented:
@jmancewicz 

- you're right about the hue, it should be a shift, not just setting it. It's not completely clear though what transformations work best in general as it heavily depends on the task (I don't know of any comprehensive study on data jittering, even on specific tasks. If you know any let me know I'm interested !). I personally had significant improvements on classification and regression tasks by modulating the hue, as well as by tinting images (by simply setting the hue). 

In fact, I think it would be even better to perform general HSV modulation (so it can be used to produce several useful transformations, see for example: http://papers.nips.cc/paper/5548-discriminative-unsupervised-feature-learning-with-convolutional-neural-networks.pdf). 

- Concerning the translation, you're right, it should be random noise or solid color.
- I also agree with changing the labels. I'm not sure about the unit of dx and dy though.

@lukeyeager yes that's a good idea, I'll do that.

lukeyeager commented:
Closing as abandoned.
